File: src/clang/magma/compile/rule/text/SuffixRule.c
#include "SuffixRule.h"
int __lambda0__(){return result;
}
magma.result.Result<magma.compile.Node, magma.compile.CompileError> parse(String input){if (!input.endsWith(suffix()))
            return new Err<>(new CompileError("Suffix '" + suffix() + "' not present", new StringContext(input)));

        String slice = input.substring(0, input.length() - suffix.length());return childRule.parse(slice);
}
magma.result.Result<String, magma.compile.CompileError> generate(magma.compile.Node node){return childRule.generate(node).mapValue(__lambda0__+suffix);
}

File: src/java/magma/SuffixRule.java
package magma;

import magma.result.Err;
import magma.result.Result;

public record SuffixRule(Rule childRule, String suffix) implements Rule {
    @Override
    public Result<String, CompileError> generate(MapNode mapNode) {
        return childRule().generate(mapNode).mapValue(inner -> inner + suffix());
    }

    @Override
    public Result<MapNode, CompileError> parse(String input) {
        if (input.endsWith(suffix)) {
            String substring = input.substring(0, input.length() - suffix.length());
            return childRule.parse(substring);
        } else {
            return new Err<>(new CompileError("Suffix '" + suffix + "' not present", new StringContext(input)));
        }
    }
}